File: backend/src/main/java/com/team05/petmeeting/domain/campaign/dto/CampaignRes.java
package com.team05.petmeeting.domain.campaign.dto;

import com.team05.petmeeting.domain.campaign.entity.Campaign;
import com.team05.petmeeting.domain.campaign.enums.CampaignStatus;

import java.util.List;

public record CampaignRes (
        int totalCampaigns,
        List<CampaignItem> campaigns
){
    public static CampaignRes of (int totalCampaigns, List<Campaign> campaignList){
        List<CampaignItem> items = campaignList.stream()
                .map(CampaignItem::from)
                .toList();
        return new CampaignRes(totalCampaigns, items);
    }
    private record CampaignItem(
        Long id,
        String title,
        String description,
        int targetAmount,
        int currentAmount,
        CampaignStatus status,
        String shelterId
    ){
        public static CampaignItem from(Campaign campaign){
            return new CampaignItem(
                    campaign.getId(),
                    campaign.getTitle(),
                    campaign.getDescription(),
                    campaign.getTargetAmount(),
                    campaign.getCurrentAmount(),
                    campaign.getStatus(),
                    campaign.getShelter().getCareRegNo()
            );
        }
    }
}

File: backend/src/main/java/com/team05/petmeeting/domain/campaign/dto/CampaignRes.kt
package com.team05.petmeeting.domain.campaign.dto

import com.team05.petmeeting.domain.campaign.entity.Campaign
import com.team05.petmeeting.domain.campaign.enums.CampaignStatus

@JvmRecord
data class CampaignRes(
    val totalCampaigns: Int,
    val campaigns: MutableList<CampaignItem?>?
) {
    @JvmRecord
    private data class CampaignItem(
        val id: Long?,
        val title: String?,
        val description: String?,
        val targetAmount: Int,
        val currentAmount: Int,
        val status: CampaignStatus?,
        val shelterId: String?
    ) {
        companion object {
            fun from(campaign: Campaign): CampaignItem {
                return CampaignItem(
                    campaign.getId(),
                    campaign.getTitle(),
                    campaign.getDescription(),
                    campaign.getTargetAmount(),
                    campaign.getCurrentAmount(),
                    campaign.getStatus(),
                    campaign.getShelter().careRegNo
                )
            }
        }
    }

    companion object {
        fun of(totalCampaigns: Int, campaignList: MutableList<Campaign?>): CampaignRes {
            val items: MutableList<CampaignItem?> = campaignList.stream()
                .map<CampaignItem?> { campaign: Campaign? -> CampaignItem.Companion.from(campaign!!) }
                .toList()
            return CampaignRes(totalCampaigns, items)
        }
    }
}
